### PR TITLE
form-qc-checker gear error reporting

### PR DESCRIPTION
**Pull request by hewanadun**
Populate detailed error metadata for validation errors and write to Flywheel input file metadata object.
Use the rule definition and error code definition schemas to map the Cerberus generated [errors](https://docs.python-cerberus.org/errors.html) with NACC QC check information. 
Schemas are stored in [nacc-qc-rules](https://s3.console.aws.amazon.com/s3/buckets/nacc-qc-rules) S3 bucket.
- Example rule definition file: [Form A2](https://s3.console.aws.amazon.com/s3/object/nacc-qc-rules?region=us-west-2&bucketType=general&prefix=UDSv4/I/rules/a2_coparticipant_demographics.json)
- Example error code definition file: [Codes A2](https://s3.console.aws.amazon.com/s3/object/nacc-qc-rules?region=us-west-2&bucketType=general&prefix=UDSv4/I/codes/a2_coparticipant_demographics.json)

NACC QC check details are retrieved from the [NACC Data Quality Checks](https://redcap.naccdata.org/redcap_v14.1.3/ProjectSetup/index.php?pid=21) REDCap project [See [report](https://redcap.naccdata.org/redcap_v14.1.3/DataExport/index.php?pid=21&report_id=8)].

See example of generated error metadata in [Flywheel]( https://naccdata.flywheel.io/#/projects/65bd1872b5708213eac0c042/sessions/65fdd2dc7c785899ddb5f37e?tab=data).

### CODE REVIEW

**Review by bjkeller — COMMENTED** [review]
This is nice. If you can do a little more to make it more readable that would be good.

File: gear/form_qc_checker/src/python/form_qc_app/error_info.py [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+                records_list = self.__redcap_con.export_records(
+                    record_ids=error_codes, fields=fields)
+                for record in records_list:
+                    try:
+                        error_desc = ErrorDescription.create(
+                            record)  # type: ignore
+                        qc_checks_list[
+                            record['error_code']] = error_desc  # type: ignore
+                    except ValidationError as error:
+                        log.error(
```
[the remark this answers is not in the document]
Comment (hewanadun) [reply gear/form_qc_checker/src/python/form_qc_app/error_info.py]: Done

File: gear/form_qc_checker/src/python/form_qc_app/error_info.py [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+                            "Error creating error description from %s: %s",
+                            record, error)
+            except REDCapConnectionError as error:
+                log.error('%s for %s', error.message, error_codes)
```
[the remark this answers is not in the document]
Comment (hewanadun) [reply gear/form_qc_checker/src/python/form_qc_app/error_info.py:162]: Not a fatal error, if NACC QC check info cannot be retrieved, the error metadata will be populated with library generated error messages.

File: gear/form_qc_checker/src/python/form_qc_app/error_info.py [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+                        error_code='qc-error',
+                        error_msg=error_msg,
+                        value=value,
+                        field=field)
+                    self.__error_writer.write(qc_error)
+                continue
+
+            code_shema = err_code_map[field]
+            valdator_errors = error_tree[field].errors
+            for error in valdator_errors:
```
[the remark this answers is not in the document]
Comment (hewanadun) [reply gear/form_qc_checker/src/python/form_qc_app/error_info.py]: Done

File: gear/form_qc_checker/src/python/form_qc_app/error_info.py [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+                # Skip if this error is a sub-error generated by a nested rule
+                # Ex. anyof, allof, etc., it will be handled at top level key
+                if error.rule != error.schema_path[1]:
+                    continue
+
+                if error.rule not in code_shema:
+                    if error.rule == Keys.NULLABLE and Keys.REQUIRED in code_shema:
```
[the remark this answers is not in the document]
Comment (hewanadun) [reply gear/form_qc_checker/src/python/form_qc_app/error_info.py]: Done

File: common/src/python/outputs/errors.py [hunk 5]
```diff
@@ -63,6 +70,24 @@ def missing_header_error() -> FileError:
                      message='No file header found')
 
 
+def system_error(
```
[the remark this answers is not in the document]
Comment (hewanadun) [reply common/src/python/outputs/errors.py:73]: Used it to report if there's something wrong with the system setup, so the validator unable to proceed. Error message will have the details on the error. Do you have any suggestions on the term to be used?

**Review by bjkeller — DISMISSED** [review]
Thanks for making those changes

**Pull request merged**